[Merged by Bors] - feat(combinatorics/simple_graph/hom): add basic definitions and lemmas for finite subgraphs

jchoules opened the pull request:
We define `finsubgraph` and its corresponding vertex- and edge-based singletons, and prove some basic ordering lemmas on the singletons. This is primarily preparatory work for proving the De Bruijn-Erdős theorem (generalised to arbitrary finite codomain graphs).

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jchoules committed:
feat(combinatorics/simple_graph/hom): add definition of finite subgraphs and prove compactness result for homomorphisms therefrom

The principal new theorem added here is `simple_graph.exists_hom_of_all_finite_homs`, the structure of whose proof is heavily cribbed from that of `finset.all_card_le_bUnion_card_iff_exists_injective` in `combinatorics/hall/basic`.

kmill commented on src/combinatorics/simple_graph/hom.lean:
```diff
@@ -0,0 +1,137 @@
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+  { G' : G.subgraph // G'.verts.finite }
```
Mathlib style is to left justify the body of definitions and theorems:
```suggestion
{ G' : G.subgraph // G'.verts.finite }
```

kmill commented on src/combinatorics/simple_graph/hom.lean:
```diff
@@ -0,0 +1,137 @@
+
+/-- A graph homomorphism from a finite subgraph of G to F. -/
+abbreviation fin_subgraph_hom (G' : G.fin_subgraph) (F : simple_graph W) := G'.val.coe →g F
+infix ` →fg ` : 50 := fin_subgraph_hom
```
Let's make this notation be local to the module.
```suggestion

local infix ` →fg ` : 50 := fin_subgraph_hom
```
When there's another use of it, we can then make this be a `localized` notation -- we generally try to keep notations from having global scope.

kmill commented on src/combinatorics/simple_graph/hom.lean:
```diff
@@ -0,0 +1,137 @@
+    adj := λ x y, (x = u ∧ y = v) ∨ (y = u ∧ x = v),
+    adj_sub := λ x y, by {rintro (⟨rfl, rfl⟩ | ⟨rfl, rfl⟩), exact e, exact e.symm},
+    edge_vert := λ x _, by {rintro (⟨rfl, _⟩ | ⟨_, rfl⟩); simp}},
+  by simp only [set.finite.insert, set.finite_singleton]⟩
```
I have a branch that defines these two graphs (not as `fin_subgraph`s though). I'll create a PR soon, merge in some of this PR, and add you as a co-author.

Naming-wise, the temporary names I came up with were `simple_graph.singleton_subgraph` and `simple_graph.subgraph_of_adj`, but your names seem good too.

kmill commented on src/combinatorics/simple_graph/hom.lean:
```diff
@@ -0,0 +1,137 @@
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
+lemma simple_graph.exists_hom_of_all_finite_homs [finite W]
+  (h : ∀ (G' : G.fin_subgraph), G' →fg F) : nonempty (G →g F) :=
```
Since we're passing actual homomorphisms to this lemma:
```suggestion
  (h : Π (G' : G.fin_subgraph), G' →fg F) : nonempty (G →g F) :=
```
This is just stylistic; it doesn't change anything.

kmill reviewed (commented):
Thanks for following up with this! I have a few comments for now.

jchoules committed:
style(combinatorics/simple_graph/hom): apply stylistic/information-hiding fixes suggested in PR

Co-authored-by: Kyle Miller <kmill31415@gmail.com>

kmill committed:
feat(combinatorics/simple_graph/subgraph): single-vertex and single-edge subgraphs

kmill committed:
lint

kmill committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

kmill committed:
Yael review plus eq_induce

kmill committed:
adj.to_walk

kmill committed:
actual tabs -> spaces

jchoules committed:
refactor(combinatorics/simple_graph): build on Kyle Miller's definitions of edge- and vertex-based singletons

jchoules committed:
Merge branch 'master' into graph_homs_compact

jchoules committed:
style(combinatorics/simple_graph/hom): bracket off separate goals arising from `refine`

jchoules committed:
refactor(combinatorics/simple_graph): rename module `hom` to `fin_subgraph`

This module has more to do with finite subgraphs than with graph homomorphisms in general.

kmill commented on src/combinatorics/simple_graph/subgraph.lean:
```diff
@@ -679,7 +679,7 @@ lemma coe_subgraph_injective (G' : G.subgraph) :
 function.left_inverse.injective restrict_coe_subgraph
 
-/-! ### Edge deletion -/
+/-! #### Edge deletion -/
```
```suggestion
/-! ### Edge deletion -/
```

kmill commented on src/combinatorics/simple_graph/subgraph.lean:
```diff
@@ -762,7 +762,7 @@ spanning_coe_le_of_le (delete_edges_le s)
 end delete_edges
 
-/-! ### Induced subgraphs -/
+/-! #### Induced subgraphs -/
```
```suggestion
/-! ### Induced subgraphs -/
```
It looks like perhaps git did something weird when merging? (Usually sectioning in a module is at level 3, but maybe you have some other intent?)

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+
+universes u v
+variables {V : Type u} {W : Type v} {G : simple_graph V} {F : simple_graph W}
+
```
Let's wrap all of your code (after or before `variables` as you prefer) in `namespace simple_graph` so that every declaration ends up in this namespace. Be sure to remove the `simple_graph` qualifiers on the two declarations that include it!

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+local infix ` →fg ` : 50 := fin_subgraph_hom
+
+/-- The finite subgraph of G generated by a single vertex. -/
+def vertex_singleton (v : V) : G.fin_subgraph := ⟨simple_graph.singleton_subgraph _ v, by simp⟩
```
Sorry to ask you to change the name of this, but for consistency I would suggest this:
```suggestion
def singleton_fin_subgraph (v : V) : G.fin_subgraph := ⟨simple_graph.singleton_subgraph _ v, by simp⟩
```

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+def vertex_singleton (v : V) : G.fin_subgraph := ⟨simple_graph.singleton_subgraph _ v, by simp⟩
+
+/-- The finite subgraph of G generated by a single edge. -/
+def edge_singleton {u v : V} (e : G.adj u v) : G.fin_subgraph :=
```
Again for consistency reasons (sorry):
```suggestion
def fin_subgraph_of_adj {u v : V} (e : G.adj u v) : G.fin_subgraph :=
```

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+
+/- Lemmas establishing the ordering between edge- and vertex-generated subgraphs. -/
+
+lemma vertex_left_le_edge {u v : V} {e : G.adj u v} : vertex_singleton u ≤ edge_singleton e :=
```
Given the previous suggestions, I might name this as
```suggestion
lemma singleton_fin_subgraph_le_left {u v : V} {e : G.adj u v} : vertex_singleton u ≤ edge_singleton e :=
```
and the next lemma as `singleton_fin_subgraph_le_right`.

These are short forms of the way-too-long names like `singleton_fin_subgraph_le_fin_subgraph_of_adj_left` (I *think* I'm putting the "left" and "right" in the correct place, but I can't say I have every mathlib naming rule internalized)

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+end
+
+/-- Given a homomorphism from a subgraph to `F`, construct its restriction to a sub-subgraph. -/
+def fin_subgraph_hom_restrict {G' G'' : G.fin_subgraph} (h : G'' ≤ G') (f : G' →fg F) : G'' →fg F :=
```
If you name it like so (assuming it's been put into the `simple_graph` namespace as suggested), then later on rather than giving the full name you can write `f.restrict h` using dot notation.
```suggestion
def fin_subgraph_hom.restrict {G' G'' : G.fin_subgraph} (h : G'' ≤ G') (f : G' →fg F) : G'' →fg F :=
```

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
+lemma simple_graph.exists_hom_of_all_finite_homs [finite W]
```
This might be a mathlib name for this lemma:
```suggestion
lemma nonempty_hom_of_forall_fin_subgraph_hom [finite W]
```
There might also be a variant for `pi` rather than `forall` since that's more accurate, but I'd need to check how we do that. In any case, `h` is morally a forall since you just need it to show all the types in the inverse system are nonempty.

(I still find this lemma to be shocking that it's true. Thanks for formalizing it!)

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+lemma vertex_left_le_edge {u v : V} {e : G.adj u v} : vertex_singleton u ≤ edge_singleton e :=
+begin
+  unfold vertex_singleton edge_singleton, simp,
+end
```
You can usually combine `unfold` and `simp`, since `unfold` uses `simp` under the hood to do its unfolding.
```suggestion
by simp [vertex_singleton, edge_singleton]
```

kmill reviewed (commented):
Once things are wrapped in a namespace and the names are mathlibified, I think I'll be happy to merge it!

YaelDillies commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+{ G' : G.subgraph // G'.verts.finite }
```
I think it would be much better to do something like
```
structure finsubgraph {V : Type u} (G : simple_graph V) :=
(verts : finset V)
(adj : V → V → Prop)
(adj_sub : ∀ {v w : V}, adj v w → G.adj v w)
(edge_vert : ∀ {v w : V}, adj v w → v ∈ verts)
(symm : symmetric adj . obviously)
```
and provide the coercion to `subgraph G` separately.

Also note that I'm changing `fin_subgraph` to `finsubgraph` to avoid confusion with `fin`.

kmill replied:
@YaelDillies For now, I was wanting to treat finite subgraphs as a construction local to the module, and I'm not sure whether we want to have a whole theory of `finsubgraph`. (I guess we should make sure the main theorem statement doesn't mention `fin_subgraph` or `->fg`.)

I can imagine a different design where we pass in `set` or `finset` as an extra argument to a generalized `subgraph` type constructor, for example, to save on *a lot* of code duplication.

YaelDillies commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
+lemma simple_graph.exists_hom_of_all_finite_homs [finite W]
```
This should still hold true, but it might strongly break your proof.
```suggestion
lemma simple_graph.exists_hom_of_all_finite_homs [countable W]
```

kmill replied:
This breaks the proof because it makes the theorem false :wink:

Having the codomain be finite is essential, and there's no generalization to arbitrary `complete_graph α`. Though the purpose of this theorem is indeed that it can be applied to `α`-colorings. For example, lifting a proof of the 4-color theorem for finite planar graphs to infinite planar graphs.

I'll leave finding counterexamples as an exercise, but let me know if you want one!

jchoules replied:
Wikipedia does suggest the existence of a generalisation of [the De Bruijn–Erdős theorem](https://en.wikipedia.org/wiki/De_Bruijn%E2%80%93Erd%C5%91s_theorem_(graph_theory)) to infinite sets of colours (for certain infinities), but it appears to require colourings to exist for more subgraphs than just the finite ones.

YaelDillies replied:
Aaah sorry I got my variables mixed up and thought `W` was the domain.

YaelDillies reviewed (commented):
Isn't there a version for homs `G → complete_graph α` (aka `α`-colorings) that doesn't require any finiteness/countability?

It would be much better if you split this PR into finite subgraphs + compactness of homs.

kmill commented on src/combinatorics/simple_graph/fin_subgraph.lean:
```diff
@@ -0,0 +1,119 @@
+  map := λ G' G'' g f, fin_subgraph_hom_restrict (category_theory.le_of_hom g.unop) f, }
+
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
```
I suppose we ought to mention what this is related to in the literature:
```suggestion
a homomorphism from the whole of `G` to `F`.

This is the graph homomorphism version of the **De Bruijn-Erdős theorem**. -/
```

jchoules committed:
style(combinatorics/simple_graph/subgraph): remove extra section indentation from botched merge

jchoules committed:
refactor(combinatorics/simple_graph/finsubgraph): encapsulate and rename module, defs, lemmas

jchoules committed:
refactor(combinatorics/simple_graph/finsubgraph): rename compactness lemma and restate it without reference to module internals

jchoules committed:
refactor(combinatorics/single_graph/finsubgraph): shorten singleton lemmas

jchoules committed:
refactor(combinatorics/simple_graph/finsubgraph): remove compactness lemma for separate pr-ing

jchoules committed:
refactor(combinatorics/simple_graph/finsubgraph): re-add compactness lemma
